Sources/AsyncAlgorithms/Locking.swift

```diff
@@ -17,100 +17,69 @@
 @_implementationOnly import WinSDK
 #endif
 
-internal struct Lock {
+private final class Lock {
 #if canImport(Darwin)
   typealias Primitive = os_unfair_lock
 #elseif canImport(Glibc)
   typealias Primitive = pthread_mutex_t
 #elseif canImport(WinSDK)
   typealias Primitive = SRWLOCK
 #endif
-  
-  typealias PlatformLock = UnsafeMutablePointer<Primitive>
-  let platformLock: PlatformLock
 
-  private init(_ platformLock: PlatformLock) {
-    self.platformLock = platformLock
-  }
-  
-  fileprivate static func initialize(_ platformLock: PlatformLock) {
+  private let _lock: UnsafeMutablePointer<Primitive>
+
+  init() {
+    _lock = UnsafeMutablePointer<Primitive>.allocate(capacity: 1)
 #if canImport(Darwin)
-    platformLock.initialize(to: os_unfair_lock())
+    _lock.initialize(to: os_unfair_lock())
 #elseif canImport(Glibc)
-    pthread_mutex_init(platformLock, nil)
+    pthread_mutex_init(_lock, nil)
 #elseif canImport(WinSDK)
-    InitializeSRWLock(platformLock)
+    InitializeSRWLock(_lock)
 #endif
   }
-  
-  fileprivate static func deinitialize(_ platformLock: PlatformLock) {
+
+  deinit {
 #if canImport(Glibc)
-    pthread_mutex_destroy(platformLock)
+    pthread_mutex_destroy(_lock)
 #endif
-    platformLock.deinitialize(count: 1)
+    _lock.deinitialize(count: 1)
+    _lock.deallocate()
   }
-  
-  fileprivate static func lock(_ platformLock: PlatformLock) {
+
+  func lock() {
 #if canImport(Darwin)
-    os_unfair_lock_lock(platformLock)
+    os_unfair_lock_lock(_lock)
 #elseif canImport(Glibc)
-    pthread_mutex_lock(platformLock)
+    pthread_mutex_lock(_lock)
 #elseif canImport(WinSDK)
-    AcquireSRWLockExclusive(platformLock)
+    AcquireSRWLockExclusive(_lock)
 #endif
   }
-  
-  fileprivate static func unlock(_ platformLock: PlatformLock) {
+
+  func unlock() {
 #if canImport(Darwin)
-    os_unfair_lock_unlock(platformLock)
+    os_unfair_lock_unlock(_lock)
 #elseif canImport(Glibc)
-    pthread_mutex_unlock(platformLock)
+    pthread_mutex_unlock(_lock)
 #elseif canImport(WinSDK)
-    ReleaseSRWLockExclusive(platformLock)
+    ReleaseSRWLockExclusive(_lock)
 #endif
   }
-
-  static func allocate() -> Lock {
-    let platformLock = PlatformLock.allocate(capacity: 1)
-    initialize(platformLock)
-    return Lock(platformLock)
-  }
-
-  func deinitialize() {
-    Lock.deinitialize(platformLock)
-  }
-
-  func lock() {
-    Lock.lock(platformLock)
-  }
-
-  func unlock() {
-    Lock.unlock(platformLock)
-  }
 }
 
-struct ManagedCriticalState<State> {
-  private final class LockedBuffer: ManagedBuffer<State, Lock.Primitive> {
-    deinit {
-      withUnsafeMutablePointerToElements { Lock.deinitialize($0) }
-    }
-  }
-  
-  private let buffer: ManagedBuffer<State, Lock.Primitive>
+final class ManagedCriticalState<State> {  
```

Comment:
This type was structural because it was using a ManagedBuffer to store both the state and the lock as properly aligned and encapsulated memory. This approach wastes a secondary allocation and has the potential of loading the memory location of the state outside of critical regions guarded by the lock.